nixos/headplane: init module, 0.6.1

igor-ramazanov opened the pull request:
Adding new packages `headplane` and `headplane-agent`, and a new service `services.headplane`.

A web UI for the `headscale`, an open-source reimplementation of the `tailscale` control server.

https://headplane.net.

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [x] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

igor-ramazanov committed:
nixos/headplane: init module

https://github.com/tale/headplane/releases/tag/v0.6.1

Co-authored-by: Erik Parawell <parawell.erik@gmail.com>

awwpotato commented on pkgs/by-name/he/headplane/package.nix:
```diff
@@ -0,0 +1,51 @@
+  stdenv,
+  ...
+}:
+stdenv.mkDerivation (finalAttrs: rec {
```
rec and finalAttrs server the same purpose  
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

awwpotato commented on pkgs/by-name/he/headplane/package.nix:
```diff
@@ -0,0 +1,51 @@
+        --add-flags $out/share/headplane/build/server/index.js
+    runHook postInstall
+  '';
+})
```
needs [meta section](https://nixos.org/manual/nixpkgs/unstable/#chap-meta)

awwpotato commented on pkgs/top-level/all-packages.nix:
```diff
@@ -4222,6 +4222,10 @@ with pkgs;
     buildGoModule = buildGo123Module;
   };
 
+  headplane = callPackage ../by-name/he/headplane { };
+
+  headplane-agent = callPackage ../by-name/he/headplane-agent { };
+
```
packages in by-name are automatically called
```suggestion
```

awwpotato commented on pkgs/by-name/he/headplane-agent/package.nix:
```diff
@@ -0,0 +1,16 @@
+{buildGoModule, fetchFromGitHub}:
+buildGoModule rec {
+  pname = "hp_agent";
+  version = "0.5.10";
+  src = fetchFromGitHub {
+    repo = "headplane";
+    owner = "tale";
+    rev = version;
```
```suggestion
buildGoModule (finalAttrs: {
  pname = "hp_agent";
  version = "0.5.10";
  src = fetchFromGitHub {
    repo = "headplane";
    owner = "tale";
    rev = finalAttrs.version;
```

awwpotato commented on pkgs/by-name/he/headplane-agent/package.nix:
```diff
@@ -0,0 +1,16 @@
+  vendorHash = "sha256-G0kahv3mPTL/mxU2U+0IytJaFVPXMbMBktbLMfM0BO8=";
+  ldflags = ["-s" "-w"];
+  env.CGO_ENABLED = 0;
+}
```
also needs meta section

igor-ramazanov replied:
Done

0xda157 reviewed (changes requested):
Please read the [contributing guide](https://nixos.org/manual/nixpkgs/unstable/#chap-meta) and [package contribution guide](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md). Also all your code needs to formatted with nixfmt-rfc-style.

awwpotato commented on pkgs/by-name/he/headplane/package.nix:
```diff
@@ -0,0 +1,60 @@
+  nodejs_22,
+  pnpm_10,
+  stdenv,
+  ...
```
```suggestion
```

awwpotato commented on pkgs/by-name/he/headplane/package.nix:
```diff
@@ -0,0 +1,60 @@
+
+  meta = {
+    description = "A feature-complete Web UI for Headscale";
+    homepage = "https://github.com/tale/headplane";
```
```suggestion
    homepage = "https://github.com/tale/headplane";
    changelog = "https://github.com/tale/headplane/releases/tag/${finalAttrs.version}";
```

awwpotato commented on pkgs/by-name/he/headplane-agent/package.nix:
```diff
@@ -0,0 +1,32 @@
+  fetchFromGitHub,
+  lib,
+}:
+let
+  version = "0.5.10";
+in
+buildGoModule {
+  pname = "hp_agent";
+  src = fetchFromGitHub {
+    repo = "headplane";
+    owner = "tale";
+    rev = version;
```
using finalAttrs will make is so overriding the version will also apply to the tag
```suggestion
buildGoModule (finalAttrs: {
  pname = "hp_agent";
  version = "0.5.10";
  
  src = fetchFromGitHub {
    repo = "headplane";
    owner = "tale";
    tag = finalAttrs.version;
```

igor-ramazanov replied:
@awwpotato Thank you for the feedback!

I think it's going to fail with something like `expected a set, got a lambda function`.

Is there any way to run CI locally? I would like to avoid pushing only to see if it passes the checks.

igor-ramazanov replied:
Yep: https://github.com/NixOS/nixpkgs/actions/runs/14474373880/job/40596271592?pr=398667

xmaso-paloit commented:
It seems that upstream [has packaged the tool](https://github.com/tale/headplane/blob/main/flake.nix#L31-L34) already. They even provide a [NixOS module](https://github.com/tale/headplane/blob/main/flake.nix#L67).

I don't 100% know what our policy is regarding this case, but it feels silly to add it to `nixpkgs`, because it adds maintenance burden here, and we will always lag behind upstream.

Pamplemousse commented:
It seems that upstream [has packaged the tool](https://github.com/tale/headplane/blob/main/flake.nix#L31-L34) already. They even provide a [NixOS module](https://github.com/tale/headplane/blob/main/flake.nix#L67).

I don't 100% know what our policy is regarding this case, but it feels silly to add it to `nixpkgs`, because it adds maintenance burden here, and we will always lag behind upstream.

igor-ramazanov commented:
@Pamplemousse It was actually me who added `nix` support there :sweat_smile:.

I thought that it's better to add to the `nixpkgs` to provide more streamlined experience, but not sure too, since I am not that familiar with the `nixpkgs`.

Pamplemousse commented on nixos/modules/services/networking/headplane.nix:
```diff
@@ -0,0 +1,108 @@
+      requires = [ "headscale.service" ];
+
+      serviceConfig = {
+        User = config.services.headscale.user;
+        Group = config.services.headscale.group;
```
You can replace `config.services.headscale` with `cfg` thanks to L19.

igor-ramazanov replied:
No, this service depends on `headscale`'s settings which is a different package: `headscale` vs. `headplane`.

Pamplemousse commented on nixos/modules/services/networking/headplane.nix:
```diff
@@ -0,0 +1,108 @@
+        User = config.services.headscale.user;
+        Group = config.services.headscale.group;
+
+        ExecStart = "${pkgs.headplane}/bin/headplane";
```
You should probably use [lib.getExe](https://nixos.org/manual/nixpkgs/stable/#function-library-lib.meta.getExe) here.

Pamplemousse commented on nixos/modules/services/networking/headplane.nix:
```diff
@@ -0,0 +1,108 @@
+      description = "Headscale Web UI";
+
+      wantedBy = [ "multi-user.target" ];
+      after = [ "headscale.service" ];
```
This probably requires `"network-online.target" `.

Pamplemousse commented on nixos/modules/services/networking/headplane.nix:
```diff
@@ -0,0 +1,108 @@
+      description = "Headplane side-running agent";
+
+      wantedBy = [ "multi-user.target" ];
+      after = [ "headplane.service" ];
```
This probably requires `"network-online.target" `.

igor-ramazanov replied:
Done

Pamplemousse commented on nixos/modules/services/networking/headplane.nix:
```diff
@@ -0,0 +1,108 @@
+        User = config.services.headscale.user;
+        Group = config.services.headscale.group;
+
+        ExecStart = "${pkgs.headplane-agent}/bin/hp_agent";
```
You should probably use [lib.getExe](https://nixos.org/manual/nixpkgs/stable/#function-library-lib.meta.getExe) here.

igor-ramazanov replied:
Thanks, done. Also used `utils.escapeSystemdExecArgs` like in other services - seems like a good practice.

Pamplemousse commented on pkgs/by-name/he/headplane/package.nix:
```diff
@@ -0,0 +1,60 @@
+    runHook preInstall
+    mkdir -p $out/{bin,share/headplane}
+    cp -r build $out/share/headplane/
+    sed -i "s;$PWD;../..;" $out/share/headplane/build/server/index.js
```
You should favor one of the substitution functions from https://nixos.org/manual/nixpkgs/stable/#ssec-stdenv-functions .

Pamplemousse commented on nixos/modules/services/networking/headplane.nix:
```diff
@@ -0,0 +1,108 @@
+
+    settings = mkOption {
+      type = types.submodule {
+        freeformType = settingsFormat.type;
```
The current configuration file doesn't seem too long. It would be nice to leverage the type system to help prevent users shooting themselves in the foot instead of a "free form" type.

Pamplemousse commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3819,6 +3819,8 @@ with pkgs;
     buildGoModule = buildGo123Module;
   };
 
+  hiera-eyaml = callPackage ../tools/system/hiera-eyaml { };
```
Not sure if it's GitHub confused with the diff, or if this has been added by mistake.

igor-ramazanov replied:
Done

StealthBadger747 commented on nixos/modules/services/networking/headplane.nix:
```diff
@@ -0,0 +1,108 @@
+          enable = mkEnableOption "headplane-agent";
+          package = mkPackageOption pkgs "headplane-agent" { };
+          settings = mkOption {
+            type = types.attrsOf [
+              types.str
+              types.bool
+            ];
```
@igor-ramazanov I've been working on adding support for headplane to interpret paths https://github.com/tale/headplane/compare/main...StealthBadger747:headplane:main

igor-ramazanov replied:
@StealthBadger747 Nice and thanks! Feel free to tag me when you a open PR. Those, better module options, then could be reused to address review comments here.

tale replied:
@StealthBadger747 open a PR sooner than later, its easier to get changes in and ship with `0.6.0` than after the fact. There's also the part where I would prefer to remove the upstream nix stuff if its not strictly necessary since it's being packaged in `nixpkgs`.

cheriimoya commented on pkgs/by-name/he/headplane/package.nix:
```diff
@@ -0,0 +1,60 @@
+  pname = "headplane";
+  version = "0.5.10";
+  src = fetchFromGitHub {
+    repo = "tale";
+    owner = "headplane";
```
```suggestion
    owner = "tale";
    repo = "headplane";
```

SuperSandro2000 commented on pkgs/by-name/he/headplane-agent/package.nix:
```diff
@@ -0,0 +1,33 @@
+  CGO_ENABLED = 0;
+
+  meta = {
+    description = "An optional sidecar process providing additional features for headplane";
```
```suggestion
    description = "Optional sidecar process providing additional features for headplane";
```

tale commented:
As the maintainer I recommend this gets updated to v0.6.1, please don't ship versions that actively have a security advisory.

igor-ramazanov commented:
@StealthBadger747 Done!

StealthBadger747 commented:
Found a few more things that either needed to be added like the module-list, or needed explanation and cleanup.

I'm done and ready for a full review

StealthBadger747 commented:
Small update. I have successfully deployed and integrated it onto a live host. Seems to work as intended.

rhoriguchi commented:
`cookie_domain` is missing in the server block. https://github.com/tale/headplane/blob/c0c4ecf6311f93e05a65eb72dfa59b5b751bfcd5/app/server/config/schema.ts#L26

rhoriguchi commented on nixos/modules/services/networking/headplane.nix:
```diff
@@ -0,0 +1,439 @@
+              options = {
+                url = mkOption {
+                  type = types.str;
+                  default = "http://127.0.0.1:8080";
```
If you run headplane you probably also run `headscale` on the same machine or at least make it conditional if headscale is enabled.

```suggestion
                  default = "http://127.0.0.1:${toString config.services.headscale.port}";
                  defaultText = lib.literalExpression ''http://127.0.0.1:''${toString config.services.headscale.port}'';
```

rhoriguchi replied:
If you run headplane, you probably also run headscale on the same machine, or at least make this conditional so that when headplane is enabled, set this as the default.

```suggestion
                  default = "http://127.0.0.1:${toString config.services.headscale.port}";
                  defaultText = lib.literalExpression ''http://127.0.0.1:''${toString config.services.headscale.port}'';
```

rhoriguchi commented on nixos/modules/services/networking/headplane.nix:
```diff
@@ -0,0 +1,439 @@
+                  example = "https://headscale.example.com";
+                };
+
+                config_path = mkOption {
```
We could do something like dnsmasq does and set that path as the default, so that as a user, you don't have to do anything. 

https://github.com/NixOS/nixpkgs/blob/5ae3b07d8d6527c42f17c876e404993199144b6a/nixos/modules/services/networking/dnsmasq.nix#L118-L123

StealthBadger747 replied:
Great PR. I'm really glad the UX is being solved for this.

Doing this was not an intuitive way of accomplishing it:

```
headplane = {
  headscale = {
    url = "https://${headscale_fqdn}";
    config_path = "${(pkgs.formats.yaml {}).generate "headscale.yml" (
      lib.recursiveUpdate
      config.services.headscale.settings
      {
        tls_cert_path = "/dev/null";
        tls_key_path = "/dev/null";
        policy.path = "/dev/null";
      }
    )}";
    config_strict = true;
  };
};
```

rhoriguchi commented on nixos/modules/services/networking/headplane.nix:
```diff
@@ -0,0 +1,439 @@
+                    to ensure that it changes the configuration in a safe way.
+                    If you want to disable this validation, set this to false.
+                  '';
+                };
```
This has to be set to `false` if https://github.com/NixOS/nixpkgs/pull/398667/files#r2566370156 used

StealthBadger747 replied:
Yep, but I'd rather get this in without modifying headscale first.

tale commented on pkgs/by-name/he/headplane/package.nix:
```diff
@@ -0,0 +1,125 @@
+    if stdenv.hostPlatform.isDarwin then
+      "sha256-H06zDYElXXiZepqxH/uCdePW6jB6VjiAX7bXnVjxwq0="
+    else
+      "sha256-KyUcaR2Lvu5kT8arr4ZO8rCa5HWXTqmk8C7P8WoYK+c=";
```
For context, this is most likely due to `@libsql/client`, I think there might be a way to get PNPM to install both platforms as part of its node modules, but not sure about the way to approach it.

SuperSandro2000 commented on pkgs/by-name/he/headplane/package.nix:
```diff
@@ -0,0 +1,125 @@
+      "i686-linux"
+      "x86_64-darwin"
+      "aarch64-darwin"
+    ];
```
```suggestion
    platforms = lib.platforms.linux + lib.platforms.darwin;
```
For our purposes this is that

Stunkymonkey commented on pkgs/by-name/he/headplane/package.nix:
```diff
@@ -0,0 +1,119 @@
+      stealthbadger747
+    ];
+    mainProgram = "headplane";
+    platforms = lib.platforms.linux + lib.platforms.darwin;
```
```suggestion
    platforms = lib.platforms.linux ++ lib.platforms.darwin;
```

Stunkymonkey commented on pkgs/by-name/he/headplane/package.nix:
```diff
@@ -0,0 +1,125 @@
+    git
+  ];
+
+  dontCheckForBrokenSymlinks = true;
```
works for me with `fetcherVersion = 2;`. so we can remove it:

```suggestion
```

igor-ramazanov committed:
maintainers: `igor-ramazanov`

igor-ramazanov committed:
maintainers: add `stealthbadger747`

igor-ramazanov committed:
headplane: init at 0.6.1

https://github.com/tale/headplane/releases/tag/v0.6.1

Co-authored-by: Erik Parawell <parawell.erik@gmail.com>

igor-ramazanov commented on pkgs/by-name/he/headplane-agent/package.nix:
```diff
@@ -0,0 +1,33 @@
+  lib,
+}:
+buildGoModule (finalAttrs: {
+  pname = "hp_agent";
```
Done

igor-ramazanov commented:
Addressed all review commits except:
1. https://github.com/NixOS/nixpkgs/pull/398667#discussion_r2589676281 - need to check whether we still need all config adjustments to satisfy `headplane`'s schema.
2. https://github.com/NixOS/nixpkgs/pull/398667#discussion_r2566370156 and https://github.com/NixOS/nixpkgs/pull/398667#discussion_r2566379936 - blocked by https://github.com/NixOS/nixpkgs/pull/465448

rhoriguchi commented on nixos/modules/services/networking/headplane.nix:
```diff
@@ -0,0 +1,428 @@
+      wants = [ "network-online.target" ];
+      after = [
+        "network-online.target"
+        "headscale.service"
```
Not sure this is better but this binds it to the services, if it changes this will fail. Thoughts?

```suggestion
        config.systemd.services.headscale.name
```

rhoriguchi reviewed (approved):
LGTM

StealthBadger747 commented:
Awesome! Thank you everyone for all the effort put into this module!

cheriimoya commented on nixos/modules/services/networking/headplane.nix:
```diff
@@ -0,0 +1,428 @@
+                  example = "/etc/headscale/config.yaml";
+                };
+
+                config_strict = mkOption {
```
maybe mkenableoption here

igor-ramazanov replied:
Done

cheriimoya commented:
otherwise really good job! would love to see this in 26.05:D

edwtjo commented:
I first want to say, great work and that this looks mostly fine! But I have two concrete concerns and one packaging question:

1. I don’t think https://github.com/tale/headplane/issues/310 is actually fixed here? The module still appears to generate an `oidc` block from defaults and only filters out null values, which matches the upstream bug report about empty/default OIDC fields being serialized. 
2. `integration.agent.executable_path` looks suspicious for 0.6.1. AFAICT upstream added that config field in 0.7.0-beta.1; the v0.6.1 config used `integration.agent.pre_authkey` instead. Also, `integration.agent.package` looks like a NixOS-internal option and probably should not live under the serialized settings tree. Please take a gander at https://github.com/tale/headplane/issues/455 to ensure it is accounted for (either explicitly or by stating it is not relevant for this nixpkg version and nixos module anymore).
3. Packaging-wise: is splitting `headplane` and `headplane-agent` reasonable? One acceptable alternative could be to have a `passthru.agent`and have an `agent.nix` under the same `headplane` directory (which also means that `src` wouldn't be duplicated since it would be passed directly to `agent.nix`. It can, however, make sense to expose both in the tree since the agent is optional and potentially be run without the UI? I'm not to familiar with the Go packaging but I believe that `headplane-agent` should, at any rate, explicitly use `subPackages = [ "cmd/hp_agent" ]`.

StealthBadger747 committed:
nixos/headplane: address review feedback

StealthBadger747 committed:
nixos/headplane: add config assertions and fix option examples

StealthBadger747 commented:
Thanks @edwtjo, addressed these.

1. The generated config now filters empty attrsets as well as nulls, and `settings.oidc` defaults to `null`, so the module no longer serializes an OIDC block unless the user explicitly configures OIDC.

2. For the agent config:

- `integration.agent.package` was only needed by the NixOS module to choose the packaged agent binary, not by Headplane itself. I moved it to `services.headplane.agent.package`, so it is no longer emitted to Headplane’s YAML config.

- I kept `integration.agent.executable_path` because it is valid for v0.6.1: the v0.6.1 tag schema accepts it here: https://github.com/tale/headplane/blob/v0.6.1/app/server/config/schema.ts#L132-L140, and the v0.6.1 server uses it when spawning the agent here: https://github.com/tale/headplane/blob/v0.6.1/app/server/hp-agent.ts#L236-L249. So this is not relying on the 0.7.0-beta.1 schema.

- Good eye on tale/headplane#455. I changed the module so `integration.agent` is not emitted at all unless the agent is enabled. If it is enabled, the module now requires `pre_authkey_path`, so we avoid generating the startup-breaking agent config.

3. I kept `headplane-agent` split out for now because upstream’s flake exposes it separately and the module needs a package path for the optional helper. That said, I think it is not very useful independently; it is a helper spawned by Headplane rather than a standalone daemon. I’m happy to switch this to `headplane.passthru.agent` if preferred. The derivation now builds only `cmd/hp_agent` via `subPackages = [ "cmd/hp_agent" ];`. 

Validation:
- Default generated YAML has no `oidc` block and no `integration.agent` block.
- Enabled-agent YAML includes `executable_path` and `pre_authkey_path`, but no `package`.
- OIDC YAML is generated only when explicitly configured.
- `headplane-agent` installs only `bin/hp_agent`.
- Built both `headplane` and `headplane-agent` on x86_64-linux.

edwtjo reviewed (approved):
I think these are good to go, some nice to haves for the future: `nixosTests.headplane` and adding `agent` under passthru

StealthBadger747 commented:
Thanks @edwtjo, @igor-ramazanov and I are really happy this finally got through. I'll be working on making a follow up PR for the v0.6.2

Since there are some changes options `mkRemovedOptionModule` might be good, but this is also so new that I don't know if it needs the bloat. If I submit a PR later today should I include those. Thoughts?

Also I'll hopefully address the tests and add agent under passthru.

tale commented:
@StealthBadger747 @igor-ramazanov glad to see this finally get through, would love if we could have a quick discussion somewhere on Headplane to figure out what I can remove/what maintenance burden can be handed off now. Appreciate the work!

StealthBadger747 commented:
@tale sure! I'll email you.
